x-pack/plugins/security_solution/server/lib/entity_analytics/asset_criticality/migrations/schedule_ecs_compliancy_migration.ts

```diff
@@ -0,0 +1,105 @@
+ * 2.0.
+ */
+
+import type { EntityAnalyticsMigrationsParams } from '../../migrations';
+import { AssetCriticalityEcsMigrationClient } from '../asset_criticality_migration_client';
+
+const TASK_TYPE = 'security-solution-ea-asset-criticality-ecs-migration';
+const TASK_ID = `${TASK_TYPE}-task-id`;
+const TASK_TIMEOUT = '15m';
+const TASK_SCOPE = ['securitySolution'];
+
+export const scheduleAssetCriticalityEcsCompliancyMigration = async ({
+  auditLogger,
+  taskManager,
+  logger,
+  getStartServices,
+}: EntityAnalyticsMigrationsParams) => {
+  if (!taskManager) {
+    return;
+  }
+
+  logger.debug(`Register task "${TASK_TYPE}"`);
+
+  taskManager.registerTaskDefinitions({
+    [TASK_TYPE]: {
+      title: `Migrate Asset Criticality index data to be ECS compliant`,
+      timeout: TASK_TIMEOUT,
+      createTaskRunner: createMigrationTask({ auditLogger, logger, getStartServices }),
+    },
+  });
+
+  const [coreStart, depsStart] = await getStartServices();
+  const taskManagerStart = depsStart.taskManager;
+  const esClient = coreStart.elasticsearch.client.asInternalUser;
+
+  const migrationClient = new AssetCriticalityEcsMigrationClient({
+    esClient,
+    logger,
+    auditLogger,
+  });
+
+  const shouldMigrateMappings = await migrationClient.isEcsMappingsMigrationRequired();
+  if (shouldMigrateMappings) {
+    logger.debug('Migrating Asset Criticality mappings');
+    await migrationClient.migrateEcsMappings();
+  }
+
+  const shouldMigrateData = await migrationClient.isEcsDataMigrationRequired();
+  if (shouldMigrateData && taskManagerStart) {
+    logger.debug(`Task scheduled: "${TASK_TYPE}"`);
+
+    const now = new Date();
+    try {
+      await taskManagerStart.ensureScheduled({
+        id: TASK_ID,
+        taskType: TASK_TYPE,
+        scheduledAt: now,
+        runAt: now,
+        scope: TASK_SCOPE,
+        params: {},
+        state: {},
+      });
+    } catch (e) {
+      logger.error(`Error scheduling ${TASK_ID}, received ${e.message}`);
+    }
+  }
+};
+
+export const createMigrationTask =
+  ({
+    getStartServices,
+    logger,
+    auditLogger,
+  }: Pick<EntityAnalyticsMigrationsParams, 'getStartServices' | 'logger' | 'auditLogger'>) =>
+  () => {
+    return {
+      run: async () => {
+        const [coreStart] = await getStartServices();
+        const esClient = coreStart.elasticsearch.client.asInternalUser;
+        const migrationClient = new AssetCriticalityEcsMigrationClient({
+          esClient,
+          logger,
+          auditLogger,
+        });
+
+        const response = await migrationClient.migrateEcsData();
+        const failures = response.failures?.map((failure) => failure.cause);
+        const hasFailures = failures && failures?.length > 0;
+
+        logger.info(
+          `Task "${TASK_TYPE}" finished. Updated documents: ${response.updated}, failures: ${
+            hasFailures ? failures.join('\n') : 0
+          }`
+        );
+      },
+
+      cancel: async () => {
```

Comment:
If this task gets cancelled, is any way to actually end it? i looks like you're using an ES client to perform the migration. Could you create an abort controller that aborts any in progress ES requests if the task gets cancelled? Otherwise this cancel function gets called but nothing actually gets cancelled.

Reply:
It makes sense to me to support cancelling, even though if the task is cancelled, the migration will fail, and some features will be disabled. 

Here is the change: https://github.com/elastic/kibana/pull/194109/commits/a9061c4e483d4a154519ddb9903a01ce894fe207